### PR TITLE
DOC: distutils: Add a docstring to show_config().

### LINKED ISSUE #9258: `show_config()` has no docstring and is not in the online docs.
The subject says it all, but in case you missed it:  the function `numpy.show_config()` (an alias for `numpy.__config__.show()`) doesn't have a docstring, and it does not appear in the [online docs](https://docs.scipy.org/doc/numpy/genindex.html#S).

A docstring that explains the information printed by `show_config()` would be helpful.

### PR DESCRIPTION
**Pull request by WarrenWeckesser**
Thanks to Sergey Kojoian for the original patch to create
the docstring.

Closes gh-9258.

### CODE REVIEW

File: numpy/distutils/misc_util.py [hunk 1]
```diff
@@ -2326,6 +2326,41 @@ def get_info(name):
...
+
+                See Also
+                --------
+                get_include : Returns the directory containing NumPy C
+                              header files.
+
+                Notes
+                -----
+                Classes specifying the information to be printed are defined
+                in the ``numpy.distutils.system_info`` module.
```
Comment (eric-wieser) [inline numpy/distutils/misc_util.py]:
```suggestion
                in the `numpy.distutils.system_info` module.
```
Should produce a link here

File: numpy/distutils/misc_util.py [hunk 2]
```diff
@@ -2326,6 +2326,41 @@ def get_info(name):
...
 
             def show():
+                """
+                Print information about various resources (libraries, library
+                directories, include directories, etc.) in the system on which
+                NumPy was built.
+
+                See Also
+                --------
+                get_include : Returns the directory containing NumPy C
```
Comment (eric-wieser) [inline numpy/distutils/misc_util.py]: I think usually see also comes last.

File: numpy/distutils/misc_util.py [hunk 3]
```diff
@@ -2326,6 +2326,41 @@ def get_info(name):
                 return g.get(name, g.get(name + "_info", {}))
 
             def show():
+                """
+                Print information about various resources (libraries, library
+                directories, include directories, etc.) in the system on which
+                NumPy was built.
```
Comment (eric-wieser) [inline numpy/distutils/misc_util.py]: Not sure if we care, but docstrings are supposed to start with a one-like description followed by a blank line, then an extended description.

If you span multiple lines, Sphinx will show only the first one in some places.

File: numpy/distutils/misc_util.py (hunk 2 above)
Comment (WarrenWeckesser) [reply numpy/distutils/misc_util.py]: `Examples` is the last section: https://numpydoc.readthedocs.io/en/latest/format.html#sections

File: numpy/distutils/misc_util.py (hunk 3 above)
Comment (WarrenWeckesser) [reply numpy/distutils/misc_util.py]: Good point. I copied this from the old PR, and only made a few edits.  In my local build of the docs, the complete sentence was displayed to the right of the name in the entry that is created in `routines.other.html`, similar to how `deprecate` is displayed at https://numpy.org/devdocs/reference/routines.other.html.

I'll do some more copy-editing and get this down to one line.

Comment (eric-wieser) [reply numpy/distutils/misc_util.py]: I might be thinking of a different problem, where Sphinx can't deal with a `.` in the first line.

File: numpy/distutils/misc_util.py [hunk 4]
```diff
@@ -2326,6 +2326,43 @@ def get_info(name):
                 return g.get(name, g.get(name + "_info", {}))
 
             def show():
+                """
+                Print information about BLAS and LAPACK libraries in use.
```
Comment (eric-wieser) [inline numpy/distutils/misc_util.py]: Or maybe

```suggestion
                Print information about libraries NumPy was built against.
```
Am I right in thinking this indicates what numpy actually uses? Or does it indicate what rebuilding numpy would use?
```

File: numpy/distutils/misc_util.py (hunk 3 above)
Comment (WarrenWeckesser) [reply numpy/distutils/misc_util.py]: In the latest commit, I added the single line "Print information about BLAS and LAPACK libraries in use." in front of what was already there.

File: numpy/distutils/misc_util.py (hunk 4 above)
Comment (WarrenWeckesser) [reply numpy/distutils/misc_util.py]: That change looks good to me.  I don't know the answer to those questions;  I was assuming it shows what is being used by the numpy that ran `show_config()`.  I'll take a look at the `system_info` module now.

Comment (eric-wieser) [reply numpy/distutils/misc_util.py]: Worth checking before talking my suggestion :)

Comment (WarrenWeckesser) [reply numpy/distutils/misc_util.py]: It looks like it searches the system for the available libraries.   Seems like a better name would be something like `show_system_info()` (which matches the underlying class that contains the code that does the work) or `show_available_library_info()`.  Anyway, I'll fix the text.

**Comment by eric-wieser** [comment]
I think we're missing something here. `np.show_config()` shows the system info where numpy was built, which is cached at build-time using the function you document here.

File: numpy/distutils/misc_util.py [hunk 5]
```diff
@@ -2326,6 +2326,43 @@ def get_info(name):
...
+                * ``libraries``: names of libraries found in the system
+                * ``library_dirs``: directories containing the libraries
+                * ``include_dirs``: directories containing library header files
+                * ``src_dirs``: directories containing library source files
+                * ``define_macros``: preprocessor macros used by
+                  ``distutils.setup``
+
+                Examples
+                --------
+                >>> np.show_config()
```
Comment (eric-wieser) [inline numpy/distutils/misc_util.py]: This is not the function that this docstring is attached to.

Comment (WarrenWeckesser) [reply numpy/distutils/misc_util.py]: `show_config` is the new name given to `show` when it is imported into the `numpy` namespace (in `numpy/__init__.py`).  I don't know why it isn't called `show_config` in `__config__.py`.

**Comment by WarrenWeckesser** [comment]
> I think we're missing something here. np.show_config() shows the system info where numpy was built, which is cached at build-time using the function you document here.

Yes, thanks for the reminder.  I lost that point for a bit.

**Comment by WarrenWeckesser** [comment]
How about

    Show libraries in the system on which NumPy was built.

for the one-line summary?

**Comment by eric-wieser** [comment]
I think:
* The function in `__config__.py` should contain that first line, and an explanation that it calls `numpy.distutils.misc_util.show_config` and a link
* The docstring you've written can stay as is, after adjusting the first line to indicate it's about the current system.

File: numpy/distutils/misc_util.py (hunk 5 above)
Comment (eric-wieser) [reply numpy/distutils/misc_util.py]: Isn't `np.show_config == np.__config__.show`? This file is not `np.__config__`.

**Comment by WarrenWeckesser** [comment]
I updated the one line summary.

I'm not sure what the other change is that you are suggesting.  `__config__.py` is the generated code, and it contains the function `__config__.show`.  This function is added to the `numpy` namespace as `show_config` with the line

    from numpy.__config__ import show as show_config

in `numpy/__init__.py`.  `show_config` is the public name of the function, which is (presumably) why `np.show_config()` is used in the docstring, even though the name of the function in the generated code is `show`.

There is no `numpy.distutils.misc_util.show_config`.

**Comment by eric-wieser** [comment]
Apologies, I did not expand the diff context enough to see this was part of a large string.

**Comment by WarrenWeckesser** [comment]
I squashed the two follow-up commits, and it looks like shippable has come back to life, so the tests have finally all passed.

@eric-wieser, I'm not sure if there are still changes that you'd like to see.

**Review by eric-wieser — APPROVED** [review]
Sorry for the confusion

**Comment by eric-wieser** [comment]
Looks good to me: https://11552-908607-gh.circle-artifacts.com/0/home/circleci/repo/doc/build/html/reference/generated/numpy.show_config.html#numpy.show_config

Shame the link didn't work out, but I'm not worried.

**Pull request merged**